components/notebook-controller/controllers/notebook_controller.go

```diff
@@ -449,6 +464,32 @@ func generateVirtualService(instance *v1beta1.Notebook) (*unstructured.Unstructu
 			"timeout": "300s",
 		},
 	}
+
+	// Create a template map for the headers section
+	// of the Istio VirtualService
+	headers := map[string]interface{}{
+		"request": map[string]interface{}{
+			"set": map[string]interface{}{},
+		},
+	}
+
+	// If key "notebooks.kubeflow.org/http-headers-request-set" is present on the notebook resource,
+	// split the literal string into keys and values and add them
+	// to the headers variable. Finally, add the headers section to
+	// the variable with the http spec.
+	if _, ok := annotations["notebooks.kubeflow.org/http-headers-request-set"]; ok && len(annotations["notebooks.kubeflow.org/http-headers-request-set"]) > 0 {
+		requestHeaders := strings.Split(annotations["notebooks.kubeflow.org/http-headers-request-set"], "\n")
+		for _, kv := range requestHeaders {
+			if len(strings.Split(kv, ": ")) == 2 {
+				k := strings.Split(kv, ": ")[0]
+				v := strings.Split(kv, ": ")[1]
+				headers["request"].(map[string]interface{})["set"].(map[string]interface{})[k] = v
```

Comment:
I think we should not let users arbitrarily add headers to the requests that pass from the ingress gateway. For example they could even override the ISTIO `userid-header`, the `X-Forwarded-Proto` so that the server assumes http instead of https etc.

I think we should reconsider the design we take here
also cc @yanniszark who has a lot of experience on security

Reply:
I believe if we change the method to `add` instead of `set` the header only gets added if it is not already present.

Reply:
If this is a security issue, I think https://github.com/kubeflow/kubeflow/issues/5588 is a much larger problem.

Reply:
@kimwnasptd I don't think this is a big problem, as the rest of the virtual service is not touched. So all the user can do is mess up the requests going to the service he already has access to, I don't think this would actually enable somebody to be able to access a notebook in another namespace (unlike the issue linked above). But I am all about learning so I'd like to know the actual dangers/problems this could have.

Reply:
@kimwnasptd To continue on your last comment below, there are a large amount of header configurations that can be used with RStudio. `X-RStudio-Root-Path` is one of course, `X-RStudio-Request` is another. However, much more can be set with headers: https://docs.rstudio.com/connect/admin/appendix/configuration/#ProxyAuth.Settings. I could be mistaken, but I thought you could also configure or use custom headers with JupyterLab as well. I will need to look at it further tomorrow morning, however, I'm sure there are other applications for which setting headers to configure them is supported and would be useful. I think with the URI rewriting, the configurable request headers, and allowing to set the application port (not included here but I do have a PR open for this), the notebook controller will be compatible with most web-applications a user could want to deploy. 

However, seeing as the issue described here is an issue that is already present, I do not think this is a blocker. Fixing the underlying issue will be a larger task and not something that will be done before the 1.3 release.

Reply:
@kimwnasptd There is a very simple solution here, just reject the annotation if it sets a restricted header name.

This should be easy to implement quickly, and I think we can start with a blacklist (which can be extended in future if issues are found).

So @kimwnasptd what headers should be blacklisted?

Reply:
I added a blacklist, so I believe this issue is now resolved.

Reply:
I took a more extensive look at the docs and I propose the following solution:

The headers that are used to configure parts of R-Studio are all non-standard headers, starting with `X-*`. Such example headers are
* `X-RStudio-Request`
* `X-RStudio-Root-Path`
* `X-Forwarded-Host`
* `X-Forwarded-Proto`
* `X-Auth-Token`

So I propose to go the other way around from blacklisting specific headers. Let's only allow the users to edit/set non-standard headers that start with `X-*`. This way we provide all the necessary configuration for R-Studio and at the same time don't allow the users to edit any header arbitrarily.

And if in the future we find users that need to specifically set the well defined headers for a Notebook we can discuss based on the specific use case on how to proceed and allow users to add more headers.

This is the case for the docs I looked up to now
https://docs.rstudio.com/ide/server-pro/access-and-security.htmlhttps://docs.rstudio.com/connect/admin/authentication/proxied/
https://docs.rstudio.com/connect/admin/appendix/configuration/#ProxyAuth.Settings

Reply:
And lastly, the controller will set accordingly the `X-RStudio-Root-Path` header if the `server-type` is `rstudio`. But the user could still be able to override this by manually specifying a different value.

This way in order for a CR to support RStuido it will only need the `notebooks.kubeflow.org/server-type` annotation, which provides a good default mechanism and extensibility to the users.

Reply:
@kimwnasptd A white-list rather than a blacklist would indeed be a possibility. However, by backing this into the controller administrators cannot easily customize these values. Also, using `X-*` for headers seems to have been deprecated int 2012: see https://tools.ietf.org/html/rfc6648. There might be a way to make this configurable, but I'm not sure this can be done on time. 

Regarding your last comment, are you suggesting that the current mechanisms added to the controller stay as they are, but the controller itself adds the annotations to the CR when `notebooks.kubeflow.org/server-type: rstudio` is present? Because I don't think this would work, as the controller will forcefully change the annotations if they are edited. I think what you are proposing, is essentially already what is being done except for that it is done in the frontend/backend rather than the controller. Users even looking into this feature are probably already aware, or capable of, figuring out how to change the virtual service spec.

Reply:
> Also, using X-* for headers seems to have been deprecated int 2012: see https://tools.ietf.org/html/rfc6648

This is a very good argument to make to the folks maintaining the R-Studio code :) 

> However, by backing this into the controller administrators cannot easily customize these values.

What extra headers do you think they would need to set aside, from the `X-*`  headers R-Studio knows/expects?

> Regarding your last comment, are you suggesting that the current mechanisms added to the controller stay as they are, but the controller itself adds the annotations to the CR when notebooks.kubeflow.org/server-type: rstudio is present?

No, the controller will not add any extra annotations if the `server-type` is present. It will only configure the vsvc accordingly.
Let me try to clarify with two examples:

CR 1:
```yaml
metadata:
  annotations:
    notebooks.kubeflow.org/server-type: rstudio
```

In this case the controller will see that the Notebook is for R-Studio so it will set the `/` url rewrite in the vsvc and also set the `X-RStudio-Root-Path` header in the vsvc for the prefix

CR 2:
```yaml
metadata:
  annotations:
    notebooks.kubeflow.org/server-type: rstudio
    notebooks.kubeflow.org/http-rewrite-uri: /some/path
```

In this case the controller will first do exactly what it would for CR 1. But then it would also detect the `notebooks.kubeflow.org/http-rewrite-uri` and use this value for the Istio rewrite, instead of the default it previously put. 

Also the same approach will be used with the `notebooks.kubeflow.org/http-headers-request-set`. If this annotation is present then the controller will use the `X-*` headers mentioned in this annotations instead

Reply:
@kimwnasptd This is now implemented, as well as a whitelist instead of a blacklist

Reply:
I think with the blacklist there is very little additional risk from this PR, and I think it's good to merge.

Regarding the server-type thing, we can do that in the future but its not necessary right now. 

Adding a whitelist would only act to reduce the usefulness of this feature for users why might have their own container images.

Reply:
> Regarding the server-type thing, we can do that in the future but its not necessary right now.

@thesuperzapper I can understand your urge to have this effort merged as soon as possible, but this is not a big feature so lets get it right from the beginning and not worry about it in the future. The functionality I described is almost there.

> Adding a whitelist would only act to reduce the usefulness of this feature for users why might have their own container images.

Could you elaborate on this? What use cases do you think we limit? With the mention mechanism, of allowing the user to set any `X-*` header they want, we still support all of the configuration options from R-Studio.